crates/polars-plan/src/plans/conversion/type_coercion/mod.rs

```diff
@@ -213,6 +213,48 @@ impl OptimizationRule for TypeCoercionRule {
                     options,
                 })
             },
+            AExpr::Function {
+                function: FunctionExpr::Append { upcast: true },
+                ref input,
+                options,
+            } => {
+                let input_schema = get_schema(lp_arena, lp_node);
```

Comment:
Do we do anything special here? Otherwise we can resolve this with he `FunctionFlags`.